streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -800,43 +788,37 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
         Objects.requireNonNull(selector, "selector can't be null");
         Objects.requireNonNull(serialized, "serialized can't be null");
         final SerializedInternal<KR, V> serializedInternal = new SerializedInternal<>(serialized);
-        final StatelessProcessorNode<K, V> graphNode = internalSelectKey(selector);
+        final ProcessorNode<K, V> graphNode = internalSelectKey(selector);
         graphNode.keyChangingOperation(true);
 
-        addGraphNode(graphNode);
-        return new KGroupedStreamImpl<>(builder,
-                                        graphNode.nodeName(),
-                                        sourceNodes,
-                                        serializedInternal.keySerde(),
-                                        serializedInternal.valueSerde(),
-                                        true,
-                                        graphNode);
+        builder.addGraphNode(this.streamsGraphNode, graphNode);
+        return new KGroupedStreamImpl<>(
+            builder,
+            graphNode.nodeName(),
+            sourceNodes,
+            serializedInternal.keySerde(),
+            serializedInternal.valueSerde(),
+            true,
+            graphNode
+        );
+
     }
 
     @Override
     public KGroupedStream<K, V> groupByKey() {
-        return groupByKey(Serialized.<K, V>with(null, null));
+        return groupByKey(Serialized.with(null, null));
     }
 
     @Override
     public KGroupedStream<K, V> groupByKey(final Serialized<K, V> serialized) {
         final SerializedInternal<K, V> serializedInternal = new SerializedInternal<>(serialized);
-
-        final StatelessProcessorNode<K, V> graphNode = new StatelessProcessorNode<>(
-            this.name,
-            null,
-            repartitionRequired,
-            Collections.<String>emptyList());
-
-        addGraphNode(graphNode);
-
         return new KGroupedStreamImpl<>(builder,
                                         this.name,
                                         sourceNodes,
                                         serializedInternal.keySerde(),
                                         serializedInternal.valueSerde(),
                                         this.repartitionRequired,
-                                        graphNode);
+                                        streamsGraphNode);
```

Comment:
Yes, it's intentional. What I found during testing is that we don't need to create a node here.

I initially had a logical node at this point, but it never rendered any details for the physical plan, as it's methods on the `KGroupedStreamImpl` that provide details for the next operation of the physical plan, thus we only need to create a new logical node when one of those operations are specified.

When I did create a new logical plan node here, it contained no details to render, so I needed to put in checks for `null` `ProcessorParameters`. Having a placeholder node was somewhat probalmatic, so instead of a "dummy" node which I found to be confusing as well, I removed creating a new node at this point, and IMHO is better off this way.

Reply:
Thanks!